### PR TITLE
Convert Clock

### LINKED ISSUE #2945: Removing usage of Clock as a class name and component name
I would like to propose that we modify the MSL to not use `Clock` as a class name. This would resolve some ambiguities for the user, and make it possible for the language design group to make `Clock` a keyword of the Modelica language.

Conversion rules would make that change somewhat easier.

### PR DESCRIPTION
**Pull request by beutlich**
Rename
* Modelica.Blocks.Sources.**Clock** to Modelica.Blocks.Sources.ContinuousClock
* Modelica.Electrical.Digital.Sources.**Clock** to Modelica.Electrical.Digital.Sources.DigitalClock

Close #2945.

@Daiwen FYI

### CODE REVIEW

File: Modelica/Blocks/Sources.mo [hunk 1]
```diff
@@ -118,12 +118,12 @@ variable <strong>y</strong> is both a variable and a connector.
 
   end BooleanExpression;
 
-  block Clock "Generate actual time signal"
+  block ContinuousClock "Generate actual time signal"
     extends Interfaces.SignalSource;
```
Comment (dietmarw) [inline Modelica/Blocks/Sources.mo]: Not really part of the PR but I wonder what the meaning of "actual" should be here. False friend from German instead of "current"? Or are there also non-actual time signals.

File: Modelica/Blocks/Sources.mo [hunk 2]
```diff
@@ -118,12 +118,12 @@ variable <strong>y</strong> is both a variable and a connector.
...
 
   equation
     y = offset + (if time < startTime then 0 else time - startTime);
-    annotation (
+    annotation (defaultComponentName="clock",
       Icon(coordinateSystem(
```
Comment (dietmarw) [inline Modelica/Blocks/Sources.mo]: I thought the reason for this PR was to not use "Clock" in MSL. Now having instances being called "clock" might lead users to also use "Clock". Maybe?

File: Modelica/Electrical/Digital.mo [hunk 3]
```diff
@@ -3894,9 +3894,9 @@ they can be used to specify the parameter, e.g., <strong>L.'0'</strong> for forc
             t_i =  time;
           end when;
           y =  if (not time>=startTime) or time >= t_i + t_width then L.'0' else L.'1';
-          annotation (Documentation(info="<html>
+          annotation (defaultComponentName="clock", Documentation(info="<html>
 <p>
```
Comment (dietmarw) [inline Modelica/Electrical/Digital.mo]: Same again. Maybe not use "clock" as default component name.

File: ModelicaTest/Tables/CombiTable2Ds.mo [hunk 4]
```diff
@@ -54,7 +54,7 @@ package CombiTable2Ds "Test models for Modelica.Blocks.Tables.CombiTable2Ds"
             3.2, 0.7, 0, 0, 0, 0, 0, 0, 0; 40, 0, 0, 0, 0, 0, 0, 0, 0, 0; 45, 0,
             0, 0, 0, 0, 0, 0, 0, 0; 50, 0, 0, 0, 0, 0, 0, 0, 0, 0], smoothness=
             Modelica.Blocks.Types.Smoothness.ContinuousDerivative));
-    Modelica.Blocks.Sources.Clock clock
+    Modelica.Blocks.Sources.ContinuousClock clock
```
Comment (dietmarw) [inline ModelicaTest/Tables/CombiTable2Ds.mo]: Try to avoid clock* names at the same time?

File: Modelica/Blocks/Sources.mo (hunk 2 above)
Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: I tried to be minimal invasive and only changed the class name while keeping all instance names.

Comment (dietmarw) [reply Modelica/Blocks/Sources.mo]: But by adding this you encourage the use of "clock". So I would not add the defaultComponentName annotation of you don't want to be to invasive and educating people to avoid "Clock" and therefore also "clock"

Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: OK, will also change/avoid `clock` as instance name and update the PR later.

Comment (henrikt-ma) [reply Modelica/Blocks/Sources.mo]: I actually think it's ok with `clock` as component name.  As soon as `Clock` becomes a keyword, this is not going to be an issue any more, and then `clock` is a very natural component name.

Comment (dietmarw) [reply Modelica/Blocks/Sources.mo]: Unfortunately to this day, Windows users still haven't grasped that Clock is not clock. So better try to help avoid this.

Comment (HansOlsson) [reply Modelica/Blocks/Sources.mo]: I also think that clock should be ok, even if we want to avoid Clock. 
We have the reverse issue with the MathBoolean blocks where the classes are named And, Or, Not - even though and, or, not are reserved words.

Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: > I also think that clock should be ok, even if we want to avoid Clock.

Yes, that is what I considered first. Now, I am in limbo how to proceed? Leave at is (as @HansOlsson approved) or rename `clock` and undo the defaultComponentName (as requested by @dietmarw and @henrikt-ma). We have draw here - two votes for either option.

Comment (dietmarw) [reply Modelica/Blocks/Sources.mo]: Maybe renaming existing models goes a bit far but I would still argue for not introducing the "defaultComponentName".

Comment (henrikt-ma) [reply Modelica/Blocks/Sources.mo]: I didn't want to rename `clock`.  I wanted to rename components called `Clock` to `clock`.

Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: > Maybe renaming existing models goes a bit far but I would still argue for not introducing the "defaultComponentName".

OK.

> I didn't want to rename `clock`. I wanted to rename components called `Clock` to `clock`.

#2945 only was on the class name. That is what I want to fix here.

**Comment by beutlich** [comment]
> 
> 
> In this case perhaps it would be better to rename the `ContinuousClock` class to `Timer` or `ElapsedTime` because it measures the elapsed time.
> 

Actually, I called it `SimulationTime` first, but noticed that it has parameters for offset and start time.

> For the digital clock, is it not simply a square wave generator for Logical 0 and 1? It's not a Digital Clock I would say. The implementation seems to just be Digital.Pulse with fewer parameters to set.

Yes, it is a periodic pulse.

File: Modelica/Blocks/Sources.mo (hunk 2 above)
Comment (henrikt-ma) [reply Modelica/Blocks/Sources.mo]: OK, so I suggested a scope creep of #2945.  Without doing all of it, what's the point of making any change at all?

Comment (HansOlsson) [reply Modelica/Blocks/Sources.mo]: To be clear: as I understand there are two examples (Modelica.Electrical.Digital.Examples.Counter3 and Modelica.Electrical.Digital.Examples.Counter) that use Clock as a component-name,

Those examples break naming conventions (components should start with lower-case letter) and I think it makes sense to have a larger PR for fixing all of them (really soon now). Whether we fix Clock in this PR - or do it together with all other names shouldn't be such a big issue, and I see a benefit of fixing all component names in those examples in one PR.

Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: Sure, we have #283 for the naming conventions (Yes, more than 9 years old.)

Comment (beutlich) [reply Modelica/Blocks/Sources.mo]: And could merge this PR already now.

File: Modelica/Blocks/Sources.mo (hunk 1 above)
Comment (christophclauss) [reply Modelica/Blocks/Sources.mo]: Since you asked me (I forgot whether I was the author): "Actual" is not correct. It should be "current". 
The result of this Clock model is, as far as I understand it, a clock signal in the sense of "time", what the watch shows. It is not in the sense of a pulsed clock signal. Therefore, I appreciate the change into ContinuousClock. It is clearer.

**Comment by beutlich** [comment]
> 
> 
> Please change "actual" to "current"

I am afraid this is beyond the scope of this PR since "actual" occurs many times in the MSL where it should read "current".

**Comment by beutlich** [comment]
> @christophclauss What is your idea?

@christophclauss What about the name "DigitalClock"?

**Comment by beutlich** [comment]
Hm, I edited #2945 to reflect the new scope.

**Comment by henrikt-ma** [comment]
> Hm, I edited #2945 to reflect the new scope.

Yes, I was totally fine with that.  I just didn't want it to be closed by accident due to GitHub keyword magic, but perhaps this is just something about closing GitHub issues by means of keywords that I don't understand…

**Comment by beutlich** [comment]
See [closing issues using keywords](https://help.github.com/en/articles/closing-issues-using-keywords).

**Comment by henrikt-ma** [comment]
> See [closing issues using keywords](https://help.github.com/en/articles/closing-issues-using-keywords).

Right, that's where I went to figure out if there was anything suggesting that merging this PR wouldn't automatically close (and here some stuff to prevent keyword closing of an issue) #2945 due to the message of this PR (which even shows a tooltip for _close_).

**Comment by beutlich** [comment]
Hm, so it is okay for you to close #2945?

**Comment by beutlich** [comment]
@christophclauss Do you prefer the name PulseWave instead of DigitalClock?

**Comment by beutlich** [comment]
No feedback from @christophclauss, but going to merge.
